### PR TITLE
[RFC] vim-patch:8.0.0252

### PR DESCRIPTION
**Pull request by janlazo**
**vim-patch:8.0.0252: not properly recognizing word characters between 128 and 255**

Problem:    Characters below 256 that are not one byte are not always
            recognized as word characters.
Solution:   Make vim_iswordc() and vim_iswordp() work the same way. Add a test
            for this. (Ozaki Kiichi)
https://github.com/vim/vim/commit/4019cf90b8657d4ab1c39744db63550f44f405a2

@jamessan suggested on gitter that I adapt `kword_test.c` to a Lua unit test but I don't know how. I'll create one in a separate commit.

### CODE REVIEW

**Comment by janlazo** [comment]
I added the test and it passes on Travis. It's ready for review.

**Comment by janlazo** [comment]
No review? I guess it's ready then.

**Comment by justinmk** [comment]
@jamessan Did you want to review this?

File: src/nvim/charset.c [hunk 1]
```diff
@@ -882,10 +879,12 @@ bool vim_iswordp(char_u *p)
 bool vim_iswordp_buf(char_u *p, buf_T *buf)
   FUNC_ATTR_PURE FUNC_ATTR_WARN_UNUSED_RESULT FUNC_ATTR_NONNULL_ALL
 {
+  int c = *p;
+
   if (MB_BYTE2LEN(*p) > 1) {
```
Comment (jamessan) [inline src/nvim/charset.c]: Why not use `c` here like in the upstream patch?

**Review by jamessan — APPROVED** [review]
It's nice to see we can easily (and more concisely) create equivalent unit tests.  Nice work!

File: src/nvim/charset.c (hunk 1 above)
Comment (janlazo) [reply src/nvim/charset.c]: I missed that. I thought it passed `p`, not `*p`. I'll fix it.

**Comment by justinmk** [comment]
quickbuild failure:

```
05:03:17,448 INFO  - # test/functional/eval/system_spec.lua @ 317: system() with output containing NULs replaces NULs by SOH characters
05:03:17,453 INFO  - not ok 914 - system() with output containing NULs replaces NULs by SOH characters
05:03:17,453 INFO  - # test/functional/eval/system_spec.lua @ 317
05:03:17,453 INFO  - # Failure message: test/functional/eval/system_spec.lua:318: Expected objects to be the same.
05:03:17,453 INFO  - # Passed in:
05:03:17,453 INFO  - # (string) '
05:03:17,453 INFO  - # 123456789
05:03:17,453 INFO  - # '
05:03:17,453 INFO  - # Expected:
05:03:17,453 INFO  - # (string) 'part1part2part3
05:03:17,453 INFO  - # '
05:03:17,453 INFO  - # stack traceback:
05:03:17,453 INFO  - # 	test/functional/eval/system_spec.lua:318: in function <test/functional/eval/system_spec.lua:317>
05:03:17,453 INFO  - # 
```

Seems unrelated, but noting it here for future reference.

File: test/unit/mbyte_spec.lua [hunk 2]
```diff
@@ -42,6 +43,15 @@ describe('mbyte', function()
     -- Sequences with more than four bytes
   end)
 
+  itp('utf_char2bytes', function()
+    local char_p = ffi.typeof('char[?]')
+    for c = 0, 0xFFFF do
+      local p = char_p(4, 0)
+      mbyte.utf_char2bytes(c, p)
+      eq(c, mbyte.utf_ptr2char(p))
+      eq(charset.vim_iswordc(c), charset.vim_iswordp(p))
```
Comment (justinmk) [inline test/unit/mbyte_spec.lua]: Nicely done!

**Comment by janlazo** [comment]
@justinmk Thanks. I included it and rebased the changes.

**Review by jamessan — APPROVED** [review]

**Pull request merged**